indexing-service/src/main/java/io/druid/indexing/common/task/ArchiveTask.java

```diff
@@ -91,12 +98,15 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
       log.info("OK to archive segment: %s", unusedSegment.getIdentifier());
     }
 
-    // Move segments
-    for (DataSegment segment : unusedSegments) {
-      final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
-      toolbox.getTaskActionClient().submit(new SegmentMetadataUpdateAction(ImmutableSet.of(archivedSegment)));
+    if (toolbox.getTaskActionClient().submit(new LockUpgradeAction(getInterval()))) {
+      // Move segments
+      for (DataSegment segment : unusedSegments) {
+        final DataSegment archivedSegment = toolbox.getDataSegmentArchiver().archive(segment);
```

Comment:
If we downgrade and upgrade then we need some form of retry in case the upgrade fails ? As of now there is no retry the task just fails if upgrade fails. Where should we put the retry mechanism - in the task or in the Upgrade Action itself like `LockAcquireAction` ? I think we have one more version Upgrade Action which will retry till it is able to upgrade and use that in case of tasks that spents a lot of time in critical section. The basic version would be still be useful in cases where we want to fail-fast for example when a hadoop-index tasklock is preempted by a realtime-index task. What do you think ?